### The <code>datetime</code> module supplies classes for manipulating dates and times
#### https://docs.python.org/3/library/datetime.html

## <code>time</code> object

In [18]:
from datetime import time

In [55]:
set_time = time(13,45,15,5)
#                ^  ^  ^ ^
#               ^   ^  ^  ^
#              hr min sec μs

# Time follows 24 hr clock
# set_time is an instance of time object

In [20]:
set_time

datetime.time(13, 45, 15, 5)

In [21]:
print(set_time)

13:45:15.000005


In [22]:
type(set_time)

datetime.time

#### Class-object attributes

In [37]:
# We can check maximum and minimum value time object can hold using class-object attributes min and max
print(f'MAX --> {time.max}')
print(f'MIN --> {time.min}')

MAX --> 23:59:59.999999
MIN --> 00:00:00


#### Regular attributes

In [23]:
set_time.hour

13

In [24]:
type(set_time.hour)

int

In [25]:
set_time.minute

45

In [26]:
set_time.second

15

In [27]:
set_time.microsecond

5

In [32]:
# We cannot over-write any attribute like this :(
set_time.hour = 7

AttributeError: attribute 'hour' of 'datetime.time' objects is not writable

#### Regular methods

In [56]:
# We need to use use replace() method on our set_time instance
set_time.replace(hour = 7)

datetime.time(7, 45, 15, 5)

In [57]:
print(set_time)

13:45:15.000005


In [58]:
# Since .replace() returns new date but doesn't change the date of instance, hence we have to assign to to same instance

set_time = set_time.replace(hour = 7)

In [59]:
print(set_time)

07:45:15.000005


## <code>date</code> class

In [38]:
from datetime import date

In [92]:
set_date = date(1998,10,1)

# Arguments must be in yyyy/mm/dd format

In [93]:
print(set_date)

1998-10-01


#### Class-object attributes

In [69]:
# We can check maximum and minimum value date object can hold using class-object attributes min and max
print(f'MAX --> {date.max}')
print(f'MIN --> {date.min}')

MAX --> 9999-12-31
MIN --> 0001-01-01


#### Class methods

In [94]:
# we can even set current date using .today() method, .today() is a class-method. Unlike regular-methods which takes
# instances as their first argument using 'self' word. Class-method takes class name as their first argument using 'class'
# word.

current_date = date.today()

In [95]:
print(current_date)

2020-08-16


#### Regular attributes

In [96]:
current_date.year

2020

In [97]:
current_date.month

8

In [98]:
current_date.day

16

#### Regular methods

In [99]:
# We can also use .replace() on dates. Here .replace() is a regular-method since it takes instances as its first argument

set_date = set_date.replace(2002,6,18)

In [100]:
print(set_date)

2002-06-18


In [101]:
# We can even convert a date in much detailed manner using .ctime() regular-method
set_date.ctime()

'Tue Jun 18 00:00:00 2002'

In [102]:
print(set_date.ctime())

Tue Jun 18 00:00:00 2002


## <code>datetime</code> class

In [70]:
from datetime import datetime

In [72]:
set_datetime = datetime(2007,5,23, 4,15,8)
#                       ̅ ̅ ̅ ̅ ̅^̅ ̅ ̅ ̅   ̅ ̅ ̅^̅ ̅ ̅
#                         date     time

In [74]:
print(set_datetime)

2007-05-23 04:15:08


#### Class-object attributes

In [91]:
print(f'MAX --> {datetime.max}')
print(f'MIN --> {datetime.min}')

MAX --> 9999-12-31 23:59:59.999999
MIN --> 0001-01-01 00:00:00


#### Regular attributes

In [77]:
set_datetime.year

2007

In [78]:
set_datetime.month

5

In [79]:
set_datetime.day

23

In [80]:
set_datetime.hour

4

In [81]:
set_datetime.minute

15

In [82]:
set_datetime.second

8

#### Regular methods

In [83]:
# .ctime()
set_datetime.ctime()

'Wed May 23 04:15:08 2007'

In [84]:
print(set_datetime.ctime())

Wed May 23 04:15:08 2007


In [85]:
# .replace()
set_datetime = set_datetime.replace(month = 8)

In [86]:
print(set_datetime.ctime())

Thu Aug 23 04:15:08 2007


In [87]:
# .date()
set_datetime.date()

datetime.date(2007, 8, 23)

In [88]:
print(set_datetime.date())

2007-08-23


In [89]:
# .time()
set_datetime.time()

datetime.time(4, 15, 8)

In [90]:
print(set_datetime.time())

04:15:08


#### Class methods

In [103]:
current_datetime = datetime.today()

In [104]:
print(current_datetime.ctime())

Sun Aug 16 17:03:29 2020


In [105]:
# perfect way to get get current date and time
print(datetime.today().ctime())

Sun Aug 16 17:05:07 2020


## Arithmetic

In [111]:
d1 = date(2017,12,1)
d2 = datetime(2018,1,1, 5,30,0)

print(d1)
print(d2)

2017-12-01
2018-01-01 05:30:00


In [112]:
# We must perform arithmetic on same objects
d2 - d1

TypeError: unsupported operand type(s) for -: 'datetime.datetime' and 'datetime.date'

### <code>datetime</code> object

In [144]:
d1 = datetime(2017,12,1, 17,15,30)
d2 = datetime(2018,1,1, 5,30,0)

print(d1.ctime())
print(d2.ctime())

Fri Dec  1 17:15:30 2017
Mon Jan  1 05:30:00 2018


In [130]:
d2 - d1

datetime.timedelta(days=30, seconds=44070)

In [131]:
print(d2 - d1)

30 days, 12:14:30


In [132]:
type(d2 - d1)

datetime.timedelta

In [133]:
result = d2 - d1

In [134]:
result.days

30

In [135]:
result.seconds

44070

30 days and 44070 seconds have passed since d1 till d2

### <code>date</code> object

In [137]:
d1 = date(2017,12,1)
d2 = date(2018,1,1)

print(d1)
print(d2)

2017-12-01
2018-01-01


In [138]:
result = d2 - d1

In [139]:
result

datetime.timedelta(days=31)

In [140]:
print(result)

31 days, 0:00:00


In [141]:
result.days

31

In [142]:
result.seconds

0

### <code>time</code> object

In [147]:
t1 = time(5,30,0)
t2 = time(17,15,30)

print(t1)
print(t2)

05:30:00
17:15:30


In [149]:
# We cannot perform arithmetic on time object
result = t2 - t1

TypeError: unsupported operand type(s) for -: 'datetime.time' and 'datetime.time'